gateway/blocks_gateway.go

```diff
@@ -0,0 +1,444 @@
+		opts.vs = vs
+		return nil
+	}
+}
+
+type BlockGatewayOption func(gwOptions *gwOptions) error
+
+func NewBlocksGateway(blockService blockservice.BlockService, opts ...BlockGatewayOption) (*BlocksGateway, error) {
+	var compiledOptions gwOptions
+	for _, o := range opts {
+		if err := o(&compiledOptions); err != nil {
+			return nil, err
+		}
+	}
+
+	// Setup the DAG services, which use the CAR block store.
+	dagService := merkledag.NewDAGService(blockService)
+
+	// Setup the UnixFS resolver.
+	fetcherConfig := bsfetcher.NewFetcherConfig(blockService)
+	fetcherConfig.PrototypeChooser = dagpb.AddSupportToChooser(func(lnk ipld.Link, lnkCtx ipld.LinkContext) (ipld.NodePrototype, error) {
+		if tlnkNd, ok := lnkCtx.LinkNode.(schema.TypedLinkNode); ok {
+			return tlnkNd.LinkTargetNodePrototype(), nil
+		}
+		return basicnode.Prototype.Any, nil
+	})
+	fetcher := fetcherConfig.WithReifier(unixfsnode.Reify)
+	r := resolver.NewBasicResolver(fetcher)
+
+	// Setup a name system so that we are able to resolve /ipns links.
+	var (
+		ns  namesys.NameSystem
+		err error
+		vs  routing.ValueStore
+	)
+
+	vs = compiledOptions.vs
+	if vs == nil {
+		vs = routinghelpers.Null{}
+	}
+
+	ns = compiledOptions.ns
+	if ns == nil {
+		ns, err = namesys.NewNameSystem(vs)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &BlocksGateway{
+		blockStore:   blockService.Blockstore(),
+		blockService: blockService,
+		dagService:   dagService,
+		resolver:     r,
+		routing:      vs,
+		namesys:      ns,
+	}, nil
+}
+
+func (api *BlocksGateway) Get(ctx context.Context, path ImmutablePath) (ContentPathMetadata, files.Node, error) {
+	md, nd, err := api.getNode(ctx, path)
+	if err != nil {
+		return md, nil, err
+	}
+
+	rootCodec := nd.Cid().Prefix().GetCodec()
+	// This covers both Raw blocks and terminal IPLD codecs like dag-cbor and dag-json
+	// Note: while only cbor, json, dag-cbor, and dag-json are currently supported by gateways this could change
+	if rootCodec != uint64(mc.DagPb) {
+		return md, files.NewBytesFile(nd.RawData()), nil
+	}
+
+	// This code path covers full graph, single file/directory, and range requests
+	f, err := ufile.NewUnixfsFile(ctx, api.dagService, nd)
+	// Note: there is an assumption here that non-UnixFS dag-pb should not be returned which is currently valid
+	if err != nil {
+		return md, nil, err
+	}
+	return md, f, nil
+}
+
+func (api *BlocksGateway) GetRange(ctx context.Context, path ImmutablePath, ranges ...GetRange) (ContentPathMetadata, files.File, error) {
+	md, nd, err := api.getNode(ctx, path)
+	if err != nil {
+		return md, nil, err
+	}
+
+	// This code path covers full graph, single file/directory, and range requests
+	n, err := ufile.NewUnixfsFile(ctx, api.dagService, nd)
+	if err != nil {
+		return md, nil, err
+	}
+	f, ok := n.(files.File)
+	if !ok {
+		return ContentPathMetadata{}, nil, NewErrorResponse(fmt.Errorf("can only do range requests on files, but did not get a file"), http.StatusBadRequest)
```

Comment:
I added a basic Go range request test. My issue with it is that we're actually testing the blocks gateway and not the gateway itself. Either way, it is useful. Maybe I should make a sharness test instead that we can later move to the compliance tests?

Reply:
Yeah, we don't really want to test the blocks gateway because it's mostly uninteresting. We return the same complete file as we do for `Get` because the blocks gateway implementation lazily loads whatever blocks are needed upon request and is the same thing we are using today.

However, when we have a different implementation of the interface that's when we're going to wish we had testing. Whatever test we have here should be usable against a new implementation (e.g. the one we end up using for bifrost-gateway).

Reply:
I removed the test from here and added as sharness in Kubo.

Reply:
@lidel I’m pretty positive that Range doesn’t work with ?format=dag-cbor (or dag-json for the matter) in 0.18.1 - I only added tests for format=raw for now. See https://github.com/ipfs/kubo/pull/9681.

Reply:
As per the 1:1 we had, I'm not adding tests for the dag-cbor, as it was never supported. Maybe in the future we will add support for it, but it's not a priority at the moment.